api/peer/list.go

```diff
@@ -43,20 +44,44 @@ type List interface {
 	Update(updates ListUpdates) error
 }
 
+// ListUpdates specifies the updates to be made to a List
+type ListUpdates struct {
+	// Additions are the identifiers that should be added to the list
+	Additions []Identifier
+
+	// Removals are the identifiers that should be removed to the list
+	Removals []Identifier
+}
+
 // ChooserList is both a Chooser and a List, useful for expressing both
 // capabilities of a single instance.
 type ChooserList interface {
 	Chooser
 	List
 }
 
-// ListUpdates specifies the updates to be made to a List
-type ListUpdates struct {
-	// Additions are the identifiers that should be added to the list
-	Additions []Identifier
+// IdentifierChooser determines the identifier of a peer for a request.
+// A Chooser may use an IdentifierChooser internally.
+// An identifier chooser is not responsible for managing the lifecycle of the
+// identified peers.
+type IdentifierChooser interface {
+	Choose(context.Context, *transport.Request) Identifier
+}
 
-	// Removals are the identifiers that should be removed to the list
-	Removals []Identifier
+// IdentifierList supports adding and removing individual peer identifiers,
+// obtaining and releasing a single subscriber.
+// An identifier list does not manage the identified peers, but is useful as a
+// utility for implementing a List.
+type IdentifierList interface {
+	Add(Identifier) Subscriber
```

Comment:
Unfortunately the subscriber interface doesn't include the "Status" information on the peer.  It only says "hey an identifier has changed".  Since the IdentifierList does not deal with Peers we'll probably need to add another API to propagate the information to the IdentifierList (I'm guessing in particular here we want the number of requests right?).

Maybe:
```
type IdentifierSubscription interface {
    NotifyStatus(IdentifierStatus)
}
type IdentifierStatus struct {
    Pending int
    ...
}
```
Not sure if we want to keep the same status struct or not (since the original status struct also has availability information).

Alternatively we could alter the IdentifierChooser.Choose function to return an on finish function (like the current Chooser).

Reply:
I guess we could upcast the Identifier though... Not sure if that's a safe practice though...